### ISSUE TITLE
Move array stride decoration to the struct fields instead of types

### ISSUE THREAD

**Issue by kvark** [gpuweb/gpuweb#960]
This is a follow-up from the call, related to #773, #910, and also directly relevant to https://github.com/gpuweb/gpuweb/issues/922#issuecomment-664654313

### Proposal

Recently, we agreed on `[[stride 16]] array<i32, 5>`, which is a decoration of the array type, similar to SPIR-V. The problem with this approach appears when we try to answer the question "is the array stride a part of the type?":
  - if it's a part of the type, then how would the assignment work between arrays of the different strides?
  - if it's not a part of the type, then why do we annotate the type?

We know that stride makes no sense for any arrays placed in the registers. It only matters when accessing a host-shareable structure. Once it's loaded into registers, the decoration is no longer relevant. Therefore, this is a proposal to put it on the fields instead:
```rust
type UniformBuffer = struct {
  [[offset 0, stride 16]]
  foo: array<i32, 5>,
};
```

### Benefits

- No need to introduce type-local decorations, which `stride` currently is the only example of, I think. The proposed `stride` is in line with the `offset` that we already have.
- Clarity in the assignment rules of arrays, with no special wording for how stride is handled
- Clarity in where the stride matters, and where it doesn't
- Less things for users to specify: only once for a field, instead of every time the type is used

### Concerns

**How do we work with pointers into host-sharable fields?**
It's not clear to me what we should be able to do with pointers, and what we shouldn't, yet. If I understand correctly, WebGPU doesn't have the equivalent of `VK_KHR_variable_pointers` and the corresponding `variablePointersStorageBuffer` feature. This means, we as WGSL implementors, have full knowledge of what each pointer is pointing to. Therefore, we can figure out, transparent to the shader writer, how to advance a pointer to the next position (by adding the stride), based on that knowledge.

**How do we deal with nested arrays?**
Given that it's not a common case, we could just say that you have to put the nested array into a separate structure:
```rust
// also a host-shareable struct
type Nested = struct {
  [[offset 0, stride 16]]
  sub_array: array<i32, 5>,
};
type UniformBuffer = struct {
  [[offset 0, stride 80]]
  foo: array<Nested, 3>,
};
```

**Comment by kainino0x** [comment]
Putting the decoration on the field doesn't prevent the issue that sometimes you can end up with a non-host-sharable type that contains strides - e.g. if you just have a local instance of that struct. So I'm not sure this is the right fix.

It makes sense to me to have it be part of the type, but it has no effect if the type is non-host-sharable.

If assignments between whole arrays (can you even do this?) are defined element-wise, then strides still don't have to be called out in the definition.

**Comment by kvark** [comment]
@kainino0x you just brought up an example "if you just have a local instance of that struct" of concern, and then followed up with a proposal to array copying that addresses it. I.e. if we say "assignment of arrays are defined element-wise", we can also say "assignment of structures are defined field-wise", which means your local copy of the struct doesn't care about the strides inside the host-shareable one you copied from.

**Comment by kainino0x** [comment]
Right, if they're defined that way (and probably will be) then it doesn't matter whether you put the array stride on the field or the type. But I think it makes at least that particular argument about copying not a very convincing one.

Honestly it seems that you're arguing (and I agree) that the layout is not part of the type, but part of the host sharable binding. Unfortunately since those bindings have complex structure there's no way to just put all the memory info in a decoration on the binding itself. Putting the decoration on a member just attaches the layout info to the parent struct type instead of the member type - doesn't extract it from the type system. If we can't extract it from the type system then it's more natural to me that it be part of the member type itself rather than the parent struct type.

**Comment by kainino0x** [comment]
Discussed on chat: https://matrix.to/#/!MFogdGJfnZLrDmgkBN:matrix.org/$Hqs95f91__-4Ok2wo0szYP27CrpNo8KlqbWOa8Mil3c?via=matrix.org&via=kyju.org&via=mozilla.org

**Comment by dneto0** [comment]
From discussion on chat:
https://matrix.to/#/!MFogdGJfnZLrDmgkBN:matrix.org/$3qF0GplLLhX_m5ojxQ6t312vKu1DQD14_hC_EAkR9oc?via=matrix.org&via=hope.net

@kvark said

> I think attaching the stride to the array type, in principle, is a fine idea.
Concerned about the details though: where we allow it and where we not, where it matters and where it doesn't, and how clear or confusing this is going to be to the users.

I agree. I suggest we need to see more proposal text before continuing to discuss.

**Comment by litherum** [comment]
When compiling to MSL, strides will be implemented by injecting additional members into the array, which means that two WGSL arrays with different strides will end up being distinct types in MSL.

**Comment by kainino0x** [comment]
You can't just inject additional members unless the stride is a multiple of the element size. You'd have to wrap the element in a struct type I think.

**Comment by grorg** [comment]
Discussed at the [2020-08-11 meeting](https://docs.google.com/document/d/1YcpsnMS7ioeaJi4qPN1rBnYDgyQxEcudjUfyKmmq4Ik/).

**Comment by kvark** [comment]
> You can't just inject additional members unless the stride is a multiple of the element size.

Can the array stride declaration hold any values? I thought it has to be a multiple of the element size, already (but I can't find this in the spec to confirm).

**Comment by kainino0x** [comment]
I don't know. Sorry, I keep talking about WGSL things I think I understand, but haven't been watching closely enough that I actually do. ^_^

**Comment by grorg** [comment]
Discussed again at the [2020-08-18 meeting](https://docs.google.com/document/d/1P7XNQ760P5yONHTkhGAlP-AA2VpggNDGMGHaGAE2tn0/edit#)

Resolution was that the `[[stride]]` decoration will go on the array type, and we'll review once more text is written (@dneto0)
